### PR TITLE
fixed issue 15721 - std.experimental.allocator dispose on interface

### PR DESCRIPTION
**Pull request by BBasile**

### CODE REVIEW

File: std/experimental/allocator/package.d [hunk 1]
```diff
@@ -1099,7 +1099,11 @@ void dispose(A, T)(auto ref A alloc, T p)
 if (is(T == class) || is(T == interface))
 {
     if (!p) return;
-    auto support = (cast(void*) p)[0 .. typeid(p).initializer.length];
+    static if (is(T==interface))
+        auto ob = cast(Object) p;
```
Comment (MetaLang) [inline std/experimental/allocator/package.d]: What if this cast returns null? I believe you can't cast C++ COM interfaces to Object (https://dlang.org/spec/interface.html#com-interfaces).

Comment (BBasile) [inline std/experimental/allocator/package.d]: I'm so used to CORBA interfaces that I forgot this crap exists.

Short answer IDK. I never use COM interface. Someone has to propose an alternative.

Comment (BBasile) [inline std/experimental/allocator/package.d]: @MEtaLang.

Actually I think it's not a problem. COM interfaces are reference counted, so trying to free one with `dispose` is a programming error. So there is no problem.

Comment (MetaLang) [inline std/experimental/allocator/package.d]: I guess if `cast(Object)` fails then try casting to `IUknown`? I'm not really sure how to deal with this case.

Comment (MetaLang) [inline std/experimental/allocator/package.d]: Ah, that works then. Might be worth putting in an assert that `is(T: IUknown)` is false.

Comment (BBasile) [inline std/experimental/allocator/package.d]: I'd even like to put a static assertion for this case but we need the expertise from someone who actually uses COM interfaces.

Comment (BBasile) [inline std/experimental/allocator/package.d]: according to file:///usr/share/dmd/html/d/spec/interface.html#com-interfaces

> It cannot be the argument of a DeleteExpression.

so the static assertion to reject this case is conform.

**Comment by BBasile** [comment]
#4022 updated

**Comment by BBasile** [comment]
changelog: `$(LI $(BUGZILLA 15721): possible free error when an interface was destroyed with experimental allocators dispose()`

**Comment by schveiguy** [comment]
You should be able to check whether an interface is COM or not by determining if it is a derivative of [core.stdc.windows.com.IUnknown](http://dlang.org/spec/interface.html#com-interfaces):

```D
if(is(T : IUnknown))
```

I would say you shouldn't even compile if it's one of those (even if it's a concrete class). Only Windows should be allocating/deallocating COM classes I think. Even in the docs it says "It cannot be the argument of a *DeleteExpression*."

**Comment by schveiguy** [comment]
LGTM, just update changelog.dd

**Comment by schveiguy** [comment]
As noted on #4023, changelog.dd update is not necessary, sorry for that request. Still LGTM.

**Comment by BBasile** [comment]
OK, I've removed my stupid attempts to update the changelog in both PR.

**Comment by BBasile** [comment]
green again.

**Comment by MetaLang** [comment]
LGTM

**Comment by schveiguy** [comment]
Auto-merge toggled on

**Comment by schveiguy** [comment]
@BBasile mind fixing that request? I'll merge it immediately.

**Comment by schveiguy** [comment]
Nevermind, I got it.

**Pull request merged**